data-prepper-plugins/kinesis-source/src/test/java/org/opensearch/dataprepper/plugins/kinesis/source/processor/KinesisRecordProcessorTest.java

```diff
@@ -0,0 +1,371 @@
+        when(kinesisSourceConfig.getCodec()).thenReturn(pluginModel);
+
+        InputCodec codec = mock(InputCodec.class);
+        when(pluginFactory.loadPlugin(eq(InputCodec.class), any())).thenReturn(codec);
+        when(kinesisStreamConfig.getCheckPointInterval()).thenReturn(CHECKPOINT_INTERVAL);
+        when(kinesisSourceConfig.getNumberOfRecordsToAccumulate()).thenReturn(NUMBER_OF_RECORDS_TO_ACCUMULATE);
+        when(kinesisSourceConfig.getStreams()).thenReturn(List.of(kinesisStreamConfig));
+        when(processRecordsInput.checkpointer()).thenReturn(checkpointer);
+        when(pluginMetrics.counterWithTags(ACKNOWLEDGEMENT_SET_CALLBACK_METRIC_NAME, KINESIS_STREAM_TAG_KEY, streamIdentifier.streamName())).thenReturn(acknowledgementSetCallbackCounter);
+    }
+
+    @Test
+    void testProcessRecordsWithoutAcknowledgementsCheckpointsEnabled()
+            throws Exception {
+        when(processRecordsInput.records()).thenReturn(createInputKinesisClientRecords());
+        when(kinesisSourceConfig.isAcknowledgments()).thenReturn(false);
+        when(kinesisStreamConfig.isEnableCheckPoint()).thenReturn(true);
+        when(acknowledgementSetManager.create(any(), any(Duration.class))).thenReturn(acknowledgementSet);
+
+        kinesisRecordProcessor = new KinesisRecordProcessor(buffer, kinesisSourceConfig, acknowledgementSetManager, pluginMetrics, pluginFactory, streamIdentifier);
+        kinesisRecordProcessor.initialize(initializationInput);
+
+        Thread.sleep(2000);
+
+        kinesisRecordProcessor.processRecords(processRecordsInput);
+
+        verify(checkpointer).checkpoint();
+        verify(buffer).writeAll(anyCollection(), anyInt());
+        verify(acknowledgementSetManager, times(0)).create(any(), any(Duration.class));
+    }
+
+    @Test
+    void testProcessRecordsWithAcknowledgementsCheckpointsEnabled()
+            throws Exception {
+        List<KinesisClientRecord> kinesisClientRecords = createInputKinesisClientRecords();
+        when(processRecordsInput.records()).thenReturn(kinesisClientRecords);
+        when(kinesisSourceConfig.isAcknowledgments()).thenReturn(true);
+        when(kinesisStreamConfig.isEnableCheckPoint()).thenReturn(false);
+        AtomicReference<Integer> numEventsAdded = new AtomicReference<>(0);
+        doAnswer(a -> {
+            numEventsAdded.getAndSet(numEventsAdded.get() + 1);
+            return null;
+        }).when(acknowledgementSet).add(any());
+
+        doAnswer(invocation -> {
+            Consumer<Boolean> consumer = invocation.getArgument(0);
+            consumer.accept(true);
+            return acknowledgementSet;
+        }).when(acknowledgementSetManager).create(any(Consumer.class), any(Duration.class));
+
+        kinesisRecordProcessor = new KinesisRecordProcessor(buffer, kinesisSourceConfig, acknowledgementSetManager, pluginMetrics, pluginFactory, streamIdentifier);
+        kinesisRecordProcessor.initialize(initializationInput);
+
+        Thread.sleep(2000);
```

Comment:
Use Awaitility instead of a sleep.

Reply:
I have modified the test to remove this.